### PR TITLE
Image processing: image arithmetic and logic tools

### PR DESCRIPTION
**Pull request by giltis**
This PR focuses exclusively on tools that are being added to facilitate image arithmetic. There are three major functions included in this PR which facilitate image processing using the VisTrails workflow manager. These tools are: 
- arithmetic_basic
- arithmetic_custom
- logic_basic

arithmetic_basic enables easy manipulation of (a) two arrays, (b) two constants, or (c) an array and a constant. The arithmetic operations included in this function are addition, subtraction, multiplication and division. The only serious restriction included as a part of the function is the generation of a ValueError if an array or constant is being divided by 0 (zero).

arithmetic_custom enables more complex arithmetic using up to 8 input arrays or constants. This function allows the user to specify a custom arithmetic expression using all of the specified input arrays and constants and any standard python operator. The available operators include:
- the arithmetic operators such as +, /, 
- the logic operators such as >, <, >=
- bitwise arithmetic operators,
- etc.

A complete list of the operators that can be used as a part of the expression is included in the documentation for the arithmetic_custom function.

logic_basic enables easy logical evaluation of (a) two arrays, (b) two constants, or (c) an array and a constant. The logical operations included in this function are:
- and
- or 
- not
- xor
- nand
- nor
- subtract

### CODE REVIEW

File: skxray/img_proc/mathops.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+                  'nand' : logical_NAND,
+                  'nor' : logical_NOR,
+                  'subtract' : logical_SUB
+                  }
+    output = logic_dict[operation](src_data1,
+                                   src_data2)
+    return output
+
+
+def logical_NAND(src_data1,
```
Comment (ericdill) [inline skxray/img_proc/mathops.py]: I would suggest that you keep the pattern consistent from the numpy by making `NAND`, `NOR` and `SUB` lowercase for these three functions

Comment (ericdill) [inline skxray/img_proc/mathops.py]: I would also suggest making the docstrings for these functions identical to their cousins in numpy:

```python
ef logical_and(x1, x2, out=None): # real signature unknown; restored from __doc__
    """
    logical_and(x1, x2[, out])
    
    Compute the truth value of x1 AND x2 element-wise.
    
    Parameters
    ----------
    x1, x2 : array_like
        Input arrays. `x1` and `x2` must be of the same shape.
    
    
    Returns
    -------
    y : {ndarray, bool}
        Boolean result with the same shape as `x1` and `x2` of the logical
        AND operation on corresponding elements of `x1` and `x2`.
    
    See Also
    --------
    logical_or, logical_not, logical_xor
    bitwise_and
    
    Examples
    --------
    >>> np.logical_and(True, False)
    False
    >>> np.logical_and([True, False], [False, False])
    array([False, False], dtype=bool)
    
    >>> x = np.arange(5)
    >>> np.logical_and(x>1, x<4)
    array([False, False,  True,  True, False], dtype=bool)
    """
    pass

def logical_not(x, out=None): # real signature unknown; restored from __doc__
    """
    logical_not(x[, out])
    
    Compute the truth value of NOT x element-wise.
    
    Parameters
    ----------
    x : array_like
        Logical NOT is applied to the elements of `x`.
    
    Returns
    -------
    y : bool or ndarray of bool
        Boolean result with the same shape as `x` of the NOT operation
        on elements of `x`.
    
    See Also
    --------
    logical_and, logical_or, logical_xor
    
    Examples
    --------
    >>> np.logical_not(3)
    False
    >>> np.logical_not([True, False, 0, 1])
    array([False,  True,  True, False], dtype=bool)
    
    >>> x = np.arange(5)
    >>> np.logical_not(x<3)
    array([False, False, False,  True,  True], dtype=bool)
    """
    pass

def logical_or(x1, x2, out=None): # real signature unknown; restored from __doc__
    """
    logical_or(x1, x2[, out])
    
    Compute the truth value of x1 OR x2 element-wise.
    
    Parameters
    ----------
    x1, x2 : array_like
        Logical OR is applied to the elements of `x1` and `x2`.
        They have to be of the same shape.
    
    Returns
    -------
    y : {ndarray, bool}
        Boolean result with the same shape as `x1` and `x2` of the logical
        OR operation on elements of `x1` and `x2`.
    
    See Also
    --------
    logical_and, logical_not, logical_xor
    bitwise_or
    
    Examples
    --------
    >>> np.logical_or(True, False)
    True
    >>> np.logical_or([True, False], [False, False])
    array([ True, False], dtype=bool)
    
    >>> x = np.arange(5)
    >>> np.logical_or(x < 1, x > 3)
    array([ True, False, False, False,  True], dtype=bool)
    """
    pass

def logical_xor(x1, x2, out=None): # real signature unknown; restored from __doc__
    """
    logical_xor(x1, x2[, out])
    
    Compute the truth value of x1 XOR x2, element-wise.
    
    Parameters
    ----------
    x1, x2 : array_like
        Logical XOR is applied to the elements of `x1` and `x2`.  They must
        be broadcastable to the same shape.
    
    Returns
    -------
    y : bool or ndarray of bool
        Boolean result of the logical XOR operation applied to the elements
        of `x1` and `x2`; the shape is determined by whether or not
        broadcasting of one or both arrays was required.
    
    See Also
    --------
    logical_and, logical_or, logical_not, bitwise_xor
    
    Examples
    --------
    >>> np.logical_xor(True, False)
    True
    >>> np.logical_xor([True, True, False, False], [True, False, True, False])
    array([False,  True,  True, False], dtype=bool)
    
    >>> x = np.arange(5)
    >>> np.logical_xor(x < 1, x > 3)
    array([ True, False, False, False,  True], dtype=bool)
    
    Simple example showing support of broadcasting
    
    >>> np.logical_xor(0, np.eye(2))
    array([[ True, False],
           [False,  True]], dtype=bool)
    """
    pass
```

File: skxray/img_proc/mathops.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+                    b %= a is equivalent to b =b % a
+            **= : raises the left operand to the power of the right operand
+                  and sets the total equal to the left operand,
+                  e.g.:
+                    b **= a is equivalent to b = b ** a
+            //= : divides the right operand into the left operand and
+                  then removes any values after the decimal point. The total
+                  is then set equal to the left operand,
+                 e.g.:
+                    b //= a is equivalent to b = b // a
```
Comment (ericdill) [inline skxray/img_proc/mathops.py]: I am unable to get any of these "Additional operators" to work

File: skxray/img_proc/mathops.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+def arithmetic_custom(expression,
+                      A,
+                      B,
+                      C=None,
+                      D=None,
+                      E=None,
+                      F=None,
+                      G=None,
+                      H=None):
```
Comment (ericdill) [inline skxray/img_proc/mathops.py]: A more generic version of this function can be written taking advantage of kwargs:
```python

def arithmetic_custom(expression, **kwargs):
    output = eval(parser.expr(expression).compile())
    return output
```

File: skxray/img_proc/mathops.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+    -------
+    result = mathops.arithmetic_custom('(A+C)/(B+D)', img_1, img_2, 2, 4)
+    """
+    output = eval(parser.expr(expression).compile())
+    return output
+
+
+def logic_basic(operation,
+                src_data1,
+                src_data2=None):
```
Comment (ericdill) [inline skxray/img_proc/mathops.py]: shorten this to `x1` and `x2`

File: skxray/img_proc/mathops.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+
+    Example
+    -------
+    result = mathops.arithmetic_custom('(A+C)/(B+D)', img_1, img_2, 2, 4)
+    """
+    output = eval(parser.expr(expression).compile())
+    return output
+
+
+def logic_basic(operation,
```
Comment (ericdill) [inline skxray/img_proc/mathops.py]: Rename this to `logical`

File: skxray/img_proc/mathops.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+                raise ValueError("This division operation will result in "
+                                 "division by a zero value. Please "
+                                 "reevaluate the denominator constant"
+                                 " (input_2).")
+
+    output = operation_dict[operation](input_1, input_2)
+    return output
+
+
+def arithmetic_custom(expression,
```
Comment (ericdill) [inline skxray/img_proc/mathops.py]: rename to `def arithmetic_expression` so that ti is more evocative of what the function is actually doing

File: skxray/img_proc/mathops.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+"""
+This module is designed to facilitate image arithmetic and logical operations
+on image data sets.
+"""
+
+import numpy as np
+import parser
+
+
+def arithmetic_basic(input_1,
```
Comment (ericdill) [inline skxray/img_proc/mathops.py]: 1. Rename to `arithmetic`
1. Change argument order so that `operation` is first
1. Rename args to `x1`, `x2`
1. Oh, and move this to `vttools/to_wrap/some_module_name.py`

File: skxray/img_proc/mathops.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+# Developed at the NSLS-II, Brookhaven National Laboratory
+# Developed by Gabriel Iltis, Oct. 2013
+"""
+This module is designed to facilitate image arithmetic and logical operations
+on image data sets.
+"""
+
+import numpy as np
+import parser
+
```
Comment (ericdill) [inline skxray/img_proc/mathops.py]: might as well import the local_* family from numpy

```python
from np import logical_and, logical_or, logical_not, logical_xor
```
and define an `__all__` that lives at the end of this module and includes all logical_* operations
```python
__all__ = [logical_and, logical_or, logical_not, logical_xor, logical_nand, logical_nor, logical_sub]
```

Comment (tacaswell) [inline skxray/img_proc/mathops.py]: `__all__` needs to be a list of strings, not the functions objects them selves.

File: skxray/img_proc/mathops.py (hunk 1 above)
Comment (giltis) [inline skxray/img_proc/mathops.py]: Function names and docstrings have now been updated.

File: skxray/img_proc/mathops.py (hunk 2 above)
Comment (giltis) [inline skxray/img_proc/mathops.py]: The equations we attempted to use to test these operators were similar to 3 += 4. Since the additional operators (+=, -=, %=, etc) inherently change the value of the left hand parameter, a constant will fail. The left hand parameter must be a variable. Thus all of these operations are expected to work in this function. Tests will be added to the test function in order to confirm, verify and validate functionality.

File: skxray/img_proc/mathops.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    >>>input_1 = [[0,0,1,0,0], [2,1,1,1,2], [2,0,1,0,2]]
+    >>>input_2 = [[0,0,0,0,0], [2,1,1,1,2], [0,0,0,0,0]]
+    result = logical_nand(input_1, input_2)
+    >>>result
+    [out]: array([[ True,  True,  True,  True,  True],
+                  [False, False, False, False, False],
+                  [ True,  True,  True,  True,  True]], dtype=bool)
+    """
+    output = logical_not(logical_and(x1,
+                                     x2))
```
Comment (tacaswell) [inline skxray/img_proc/mathops.py]: why the line break in all of these?

File: skxray/img_proc/mathops.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    """
+    This function enables the computation of the LOGICAL_NAND of two image or 
+    volume data sets. This function enables easy isolation of all data points 
+    NOT INCLUDED IN BOTH SOURCE DATA SETS. This function can be used for data 
+    comparison, material isolation, noise removal, or mask 
+    application/generation.
+
+    Parameters
+    ----------
+    x1 : {array, list, tuple, int, float}
```
Comment (tacaswell) [inline skxray/img_proc/mathops.py]: This should not have `{}` which indicate that the input as as enum

File: skxray/img_proc/mathops.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+    """
+    temp = logical_not(logical_and(x1,
+                                   x2))
+    output = logical_and(x1,
+                         temp)
+    return output
+
+
+__all__ = (add, subtract, multiply, divide, logical_and, logical_or,
```
Comment (tacaswell) [inline skxray/img_proc/mathops.py]: This needs to be a list of strings.

File: skxray/img_proc/mathops.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+currently provided in numpy in order to provide a complete set of logical
+operations for data analysis.
+
+The new functions include:
+    logical_nand: Identifies all elements NOT included in BOTH inputs
+    logical_nor: Identifies all elements NOT included in EITHER input
+    logical_sub: Identifies all elements ONLY included in input_1
+"""
+
+import numpy as np
```
Comment (tacaswell) [inline skxray/img_proc/mathops.py]: This isn't used and this module is missing the future-proofing imports.

File: skxray/img_proc/mathops.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+from numpy import (logical_and, logical_or, logical_not, logical_xor, add,
+                   subtract, multiply, divide)
+
+def logical_nand(x1,
```
Comment (tacaswell) [inline skxray/img_proc/mathops.py]: Again, why the line breaks?

File: skxray/img_proc/mathops.py (hunk 10 above)
Comment (tacaswell) [inline skxray/img_proc/mathops.py]: I would mirror the numpy logical_* doc strings here

File: skxray/img_proc/mathops.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+                 x2):
+    """
+    This function enables the computation of the LOGICAL_NAND of two image or
```
Comment (tacaswell) [inline skxray/img_proc/mathops.py]: This should be terser.

```python
Computes the truth value of NOT (x1 AND x2) element wise.
```

The longer narrative should go in the rst-based prose documentation.

File: skxray/img_proc/mathops.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+    """
+    output = logical_not(logical_or(x1,
+                                    x2))
+    return output
+
+
+def logical_sub(x1,
+                x2):
+    """
+    This function enables LOGICAL SUBTRACTION of one binary image or volume data
```
Comment (tacaswell) [inline skxray/img_proc/mathops.py]:
```python
Compute truth value of x1 AND (NOT ( x1 AND x2)) element wise.
```

File: skxray/tests/test_img_proc.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+    test_constant_int = 5
+    test_constant_flt = 2.0
+
+    #Int array and int constant
+    add_check = test_array_1 + test_constant_int
+    sub_check = np.subtract(test_array_1, test_constant_int)
+    mult_check = np.multiply(test_array_1, test_constant_int)
+    div_check = np.divide(test_array_1, test_constant_int)
+
+    assert_equal(mathops.arithmetic_basic(test_array_1,
```
Comment (tacaswell) [inline skxray/tests/test_img_proc.py]: Does this function still exist to be tested?

File: skxray/img_proc/mathops.py (hunk 2 above)
Comment (giltis) [inline skxray/img_proc/mathops.py]: Reference to these operators have been removed from the docstring for arithmetic_custom. A note has been included in the commit message stating that modifying the parsing function to properly handle these operators is possible, but beyond the scope of our current focus. These functions are extraneous since the operations can still be performed (e.g. the expression a=a+b is equivalent to a+=b)

File: skxray/img_proc/mathops.py (hunk 4 above)
Comment (giltis) [inline skxray/img_proc/mathops.py]: Completed for both arithmetic_basic, and logic_basic. In addition, the arg order for all of the functions has been standardized to start with the operation or expression definition, followed by the input data.

File: skxray/img_proc/mathops.py (hunk 7 above)
Comment (giltis) [inline skxray/img_proc/mathops.py]: The functions logic_basic, arithmetic_basic and arithmetic_custom have been renamed as logic, arithmetic, and arithmetic_expression as we discussed. Bullets 2,3,and 4 were addressed in previous commits and comments.

All is good in the world. Well, my world at least....

File: skxray/img_proc/mathops.py (hunk 8 above)
Comment (giltis) [inline skxray/img_proc/mathops.py]: These changes have now been made in skxray/mathops.py

File: skxray/img_proc/mathops.py (hunk 9 above)
Comment (giltis) [inline skxray/img_proc/mathops.py]: You mean the >>> ? I saw that format in other function examples, including some of yours (I believe) and figured that it was a proper way to document usage examples in docstrings. Should these simply be removed?

Comment (tacaswell) [inline skxray/img_proc/mathops.py]: No, I mean I think this line should read

```python
output = logical_notlogical_and(x1, x2))
```
The line break makes it harder to read.

Comment (giltis) [inline skxray/img_proc/mathops.py]: Oh, Okay. To me it seems easier to read. But, I've no problem in adjusting it.

File: skxray/img_proc/mathops.py (hunk 10 above)
Comment (giltis) [inline skxray/img_proc/mathops.py]: Fixed to mirror numpy docstrings.
